项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入NumPy和Pandas数据库，用Pandas的`read_csv`函数读取原始数据集'e_commerce.csv'，使其转换成DataFrame格式，并赋值给变量`df`。

展示数据集的前5行和后5行。

In [1]:
import numpy as np
import pandas as pd

In [2]:
df= pd.read_csv('e_commerce.csv')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## 评估数据

在这一部分，我将对上一部分所建立的`df`这个DataFrame所包含的数据进行评估

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 一、评估数据整洁度

利用DataFrame的info方法和sample方法，分别提供数据的概况信息和实际数据，来评估数据的整洁度

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


列名符合每列只能表示一个变量的整洁性规则

In [4]:
df.sample(5)

InvoiceNo StockCode                     Description  Quantity  \
270141    560547     82600           NO SINGING METAL SIGN        12   
480057    577252     22154     ANGEL DECORATION 3 BUTTONS         48   
250351    559023     22489  PACK OF 12 TRADITIONAL CRAYONS       216   
280101    561371    90214S       LETTER "S" BLING KEY RING         2   
161782    550488     85094    CANDY SPOT EGG WARMER RABBIT        12   

             InvoiceDate  UnitPrice  CustomerID         Country  
270141   7/19/2011 12:39       2.10     17419.0  United Kingdom  
480057  11/18/2011 12:22       0.42     18034.0  United Kingdom  
250351    7/5/2011 13:19       0.42     18142.0  United Kingdom  
280101   7/26/2011 16:25       0.83         NaN  United Kingdom  
161782   4/18/2011 14:27       0.19     16729.0  United Kingdom

从抽样的10行数据来看，符合'每列是一个变量，每行是一个观察值，每个单元格是一个值'这三个标准。具体来看每行是关于某商品的一次交易数据，每列是关于交易的各个变量，因此不存在结构性问题。

### 二、评估数据干净度

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值。

`Description`和`CustomerID`列存在缺失值。

`CustomerID`列数据类型应为object，`InvoiceDate`数据类型应为datetime，应当进行数据类型转换

#### (一)、评估丢失数据

初步统计各列缺失值的数量

In [6]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

在了解`Description`存在缺失值后，根据条件提取出缺失观察值。

In [7]:
df[df.Description.isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失`Description`变量值

从输出结果来看，这些缺失`Description`的交易数据，`UnitPrice`都为0，且`CustomerID`同时缺失。为了验证猜想，我们增加筛选条件，看是否存在`Description`变量缺失，且`UnitPrice`不为0的数据

In [8]:
df[(df.Description.isnull()) & (df.UnitPrice != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

验证猜想，输出结果为0条，说明缺失`Description`变量值的数据，同时也不具备有效的`UnitPrice`值。

`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续数据分析的关键变量。如果他们同时缺失，我们认为这些数据无法提供有效含义，因此这些后续可以被删除。

变量`CustomerID`同样存在缺失值，因此也根据条件提取出缺失观察值。

In [9]:
df[df.CustomerID.isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

有135080条交易数据缺失`CustomerID`变量值

从输出结果来看，缺失`CustomerID`变量值不一定缺失其他变量值，并且`CustomerID`不是后续分析畅销商品的关键变量，因此保留此变量为空的观察值。

#### (二)、评估重复数据

根据数据变量含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但是一次交易可以包含多件商品，因此`InvoiceNo`可以存在重复；不同交易可以选择同件商品，因此`InvoiceNo`可以存在重复；每个客户可以产生不同交易，因此`CustomerID`可以存在重复。

但是数据所有变量同时重复时，应被看作重复数据。而不是某个顾客一次买了多件同一商品，毕竟这样应该是quantity变量的值为多个。

In [10]:
df[df.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

从输出结果来看，数据中存在5268条重复数据，应在数据清洗环节删除所有变量同时重复的数据。

#### (三)、 评估不一致数据 

从数据变量含义来看，只能判断`Country`变量是否存在不一致数据

In [11]:
df.Country.value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

从`Country`变量值来看，此处U.K., UK和United Kingdom都表示英国，United States和USA都表示美国。这些数据是不一致数据，在后续清理步骤需要把数值都替换成统一的

#### (四)、评估无效/错误数据

通过DataFrame的`describe`方法，对数值统计信息进行快速了解

In [12]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`的值存在负数，这些均为无效数据。

接下来先筛选出`Quantity`为负数的观察值，来进一步观察数据的特点

In [13]:
df[df['Quantity'] <= 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

可以发现`Quantity`为负数时，似乎`InvoiceNo`均为C开头，表示这笔交易被取消。

为了验证猜想，我们需要增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

In [14]:
df[(df['Quantity'] <= 0) & (df['InvoiceNo'].str[0] != 'C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

可以发现猜想并不准确，确实存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但进一步观察发现，这些筛选出的观察值UnitPrice均为0，因此我们增加筛选条件，进一步验证

In [15]:
df[(df['Quantity'] <= 0) & (df['InvoiceNo'].str[0] != 'C') & (df['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，证明当`Quantity`为负数的时候，观察值满足以下条件之一：

`InvoiceNo`以C开头，表示订单被取消；

`UnitPrice`为0。说明单价0英镑

这些交易数据不是有效交易数据，不在后续数据分析范围内，因此在数据清理阶段，将`Quantity`为负数的观察值删除

同时，我们需要确认，`InvoiceNo`以C开头时，或`UnitPrice`为0时，`Quantity`变量是否都为负数。

这样可以帮我们确认，在数据清理环节，是否只需要删除`Quantity`为负数的数据

In [16]:
df[(df['InvoiceNo'].str[0] == 'C') & (df['Quantity'] > 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，`InvoiceNo`以C开头时，`Quantity`变量的值都大于0。因此在数据清理阶段，将`Quantity`为负数的观察值删除，就可以删除所有订单被取消的数据。

因为在评估缺失数据时，发现了一批`Description`为空值，同时`UnitPrice`为0的数据。因此在筛选时还要加上条件，`Description`不为空值

In [17]:
df[(df['UnitPrice'] == 0) & (df['Quantity'] > 0) & (df['Description'].notnull())]

InvoiceNo StockCode                       Description  Quantity  \
6391      536941     22734                            amazon        20   
6392      536942     22139                            amazon        15   
9302      537197     22841      ROUND CAKE TIN VINTAGE GREEN         1   
14335     537534     85064      CREAM SWEETHEART LETTER RACK         1   
14336     537534     84832  ZINC WILLIE WINKIE  CANDLE STICK         1   
...          ...       ...                               ...       ...   
524622    580609     22927                            Amazon         1   
535325    581202     23404                             check        41   
535334    581211     22142                             check        14   
538504    581406    46000M      POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S      POLYESTER FILLER PAD 40x40cm       300   

            InvoiceDate  UnitPrice  CustomerID         Country  
6391    12/3/2010 12:08        0.0         NaN  United Kingdom  
6392    12/3/2010 12:08        0.0         NaN  United Kingdom  
9302    12/5/2010 14:02        0.0     12647.0         Germany  
14335   12/7/2010 11:48        0.0         NaN  United Kingdom  
14336   12/7/2010 11:48        0.0         NaN  United Kingdom  
...                 ...        ...         ...             ...  
524622  12/5/2011 11:41        0.0         NaN  United Kingdom  
535325  12/7/2011 18:30        0.0         NaN  United Kingdom  
535334  12/7/2011 18:36        0.0         NaN  United Kingdom  
538504  12/8/2011 13:58        0.0         NaN  United Kingdom  
538505  12/8/2011 13:58        0.0         NaN  United Kingdom  

[587 rows x 8 columns]

根据输出结果，存在`UnitPrice`为0时，`Quantity`大于0，且`Description`不为空值的数据。

但进一步观察发现，这些筛选出的观察值，大部分`CustomerID`为NaN，因此我们增加筛选条件，进一步验证

In [18]:
df[(df['UnitPrice'] == 0) & (df['Quantity'] > 0) & (df['Description'].notnull()) & (df['CustomerID'].notnull())]

InvoiceNo StockCode                          Description  Quantity  \
9302      537197     22841         ROUND CAKE TIN VINTAGE GREEN         1   
33576     539263     22580         ADVENT CALENDAR GINGHAM SACK         4   
40089     539722     22423             REGENCY CAKESTAND 3 TIER        10   
47068     540372     22090              PAPER BUNTING RETROSPOT        24   
47070     540372     22553               PLASTERS IN TIN SKULLS        24   
56674     541109     22168        ORGANISER WOOD ANTIQUE WHITE          1   
86789     543599    84535B         FAIRY CAKES NOTEBOOK A6 SIZE        16   
130188    547417     22062  CERAMIC BOWL WITH LOVE HEART DESIGN        36   
139453    548318     22055   MINI CAKE STAND  HANGING STRAWBERY         5   
145208    548871     22162          HEART GARLAND RUSTIC PADDED         2   
157042    550188     22636   CHILDS BREAKFAST SET CIRCUS PARADE         1   
187613    553000     47566                        PARTY BUNTING         4   
198383    554037     22619            SET OF 6 SOLDIER SKITTLES        80   
279324    561284     22167           OVAL WALL MIRROR DIAMANTE          1   
282912    561669     22960             JAM MAKING SET WITH JARS        11   
285657    561916         M                               Manual         1   
298054    562973     23157           SET OF 6 NATIVITY MAGNETS        240   
314745    564651     23270     SET OF 2 CERAMIC PAINTED HEARTS         96   
314746    564651     23268  SET OF 2 CERAMIC CHRISTMAS REINDEER       192   
314747    564651     22955             36 FOIL STAR CAKE CASES        144   
314748    564651     21786                   POLKADOT RAIN HAT        144   
358655    568158      PADS           PADS TO MATCH ALL CUSHIONS         1   
361825    568384         M                               Manual         1   
379913    569716     22778                   GLASS CLOCHE SMALL         2   
395529    571035         M                               Manual         1   
420404    572893     21208          PASTEL COLOUR HONEYCOMB FAN         5   
436428    574138     23234        BISCUIT TIN VINTAGE CHRISTMAS       216   
436597    574175     22065       CHRISTMAS PUDDING TRINKET POT         12   
436961    574252         M                               Manual         1   
439361    574469     22385            JUMBO BAG SPACEBOY DESIGN        12   
446125    574879     22625                   RED KITCHEN SCALES         2   
446793    574920     22899         CHILDREN'S APRON DOLLY GIRL          1   
446794    574920     23480       MINI LIGHTS WOODLAND MUSHROOMS         1   
454463    575579     22437        SET OF 9 BLACK SKULL BALLOONS        20   
454464    575579     22089        PAPER BUNTING VINTAGE PAISLEY        24   
479079    577129     22464          HANGING METAL HEART LANTERN         4   
479546    577168         M                               Manual         1   
480649    577314     23407       SET OF 2 TRAYS HOME SWEET HOME         2   
485985    577696         M                               Manual         1   
502122    578841     84826       ASSTD DESIGN 3D PAPER STICKERS     12540   

             InvoiceDate  UnitPrice  CustomerID         Country  
9302     12/5/2010 14:02        0.0     12647.0         Germany  
33576   12/16/2010 14:36        0.0     16560.0  United Kingdom  
40089   12/21/2010 13:45        0.0     14911.0            EIRE  
47068     1/6/2011 16:41        0.0     13081.0  United Kingdom  
47070     1/6/2011 16:41        0.0     13081.0  United Kingdom  
56674    1/13/2011 15:10        0.0     15107.0  United Kingdom  
86789    2/10/2011 13:08        0.0     17560.0  United Kingdom  
130188   3/23/2011 10:25        0.0     13239.0  United Kingdom  
139453   3/30/2011 12:45        0.0     13113.0  United Kingdom  
145208    4/4/2011 14:42        0.0     14410.0  United Kingdom  
157042   4/14/2011 18:57        0.0     12457.0     Switzerland  
187613   5/12/2011 15:21        0.0     17667.0  United K

根据输出结果，猜想得到验证，证明当`UnitPrice`为0时，`Quantity`大于0，且`Description`不为空值时的数据，大部分`CustomerID`为空值，小部分不为空值。

但由于数据分析的最终目的是，提高营收，因此在数据清理环节，删除`UnitPrice`为0的数据。

接下来筛选出`UnitPrice`为负数的观察值，观察数据的特点

In [19]:
df[df['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调整，不属于实际交易的数据，因此在数据清洗环节也将其删除

通过DataFrame的`value_counts`方法，分别检查`InvoiceNo`、`StockCode`、`InvoiceDate`变量是否存在可被发现的无效/错误数据

In [20]:
df.InvoiceNo.value_counts()

573585     1114
581219      749
581492      731
580729      721
558475      705
           ... 
554023        1
554022        1
554021        1
554020        1
C558901       1
Name: InvoiceNo, Length: 25900, dtype: int64

In [21]:
df.StockCode.value_counts()

85123A    2313
22423     2203
85099B    2159
47566     1727
20725     1639
          ... 
21431        1
22275        1
17001        1
90187A       1
72759        1
Name: StockCode, Length: 4070, dtype: int64

In [22]:
df.InvoiceDate.value_counts()

10/31/2011 14:41    1114
12/8/2011 9:28       749
12/9/2011 10:03      731
12/5/2011 17:24      721
6/29/2011 15:58      705
                    ... 
9/28/2011 9:57         1
9/28/2011 9:45         1
3/16/2011 12:13        1
3/16/2011 12:22        1
3/16/2011 15:58        1
Name: InvoiceDate, Length: 23260, dtype: int64

从输出结果来看，不存在未被发现过的无效/错误数据。

## 清理数据

根据前面数据评估部分的结果，我们需要进行的数据清理包括：

- 把`CustomerID`变量的数据类型转换为字符串类型
- 把`InvoiceDate`变量的数据类型转换为日期时间类型
- 把`Description`变量存在缺失的观察值删除
- 删除所有变量同时重复的数据
- 把`Country`变量值`'U.K.'`、`'UK'`替换为`'United Kindom'`
- 把`Country`变量值`'USA'`替换为`'United States'`
- 把`Quantity`小于等于0的观察值删除
- 把`UnitPrice`观察值小于等于0的数据删除

为了区分经过清理的数据和原始的数据，我们创建新的变量`cleaned_df`。之后的清理步骤都将被运用在`cleaned_df`上。

In [23]:
cleaned_df = df
cleaned_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

把`CustomerID`变量的数据类型转换为字符串类型：

In [24]:
cleaned_df['CustomerID'] = cleaned_df['CustomerID'].astype(str)
cleaned_df['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustomerID`结尾的`".0"`删除：

In [25]:
cleaned_df['CustomerID'] = cleaned_df['CustomerID'].str.slice(0, -2)
cleaned_df['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`InvoiceDate`变量的数据类型转换为日期时间类型：

In [26]:
cleaned_df['InvoiceDate'] = pd.to_datetime(cleaned_df['InvoiceDate'])
cleaned_df.InvoiceDate

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`Description`变量存在缺失的观察值删除：

In [27]:
cleaned_df.dropna(subset=['Description'], inplace=True)
cleaned_df.Description.isnull().sum()

0

删除所有变量同时重复的数据：

In [28]:
cleaned_df.drop_duplicates(inplace=True)
cleaned_df[cleaned_df.duplicated()]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

把`Country`变量值`'U.K.'`、`'UK'`替换为`'United Kindom'`;
把`Country`变量值`'USA'`替换为`'United States'`：

In [29]:
cleaned_df.replace(['U.K.', 'UK'], 'United Kingdom', inplace=True)
cleaned_df.replace('USA', 'United States', inplace=True)
print(len(cleaned_df[cleaned_df['Country'] == 'U.K.']))
print(len(cleaned_df[cleaned_df['Country'] == 'UK']))
print(len(cleaned_df[cleaned_df['Country'] == 'USA']))

0
0
0


把`Quantity`小于等于0的观察值删除：

In [30]:
cleaned_df.drop(cleaned_df[cleaned_df.Quantity <= 0].index, inplace=True)
len(cleaned_df[cleaned_df.Quantity <= 0])

0

把`UnitPrice`观察值小于等于0的数据删除：

In [31]:
cleaned_df.drop(cleaned_df[cleaned_df.UnitPrice <= 0].index, inplace=True)
len(cleaned_df[cleaned_df.UnitPrice <= 0])

0

发现单价存在0.001英镑的，不合常理，应为错误数据

In [32]:
cleaned_df.drop(cleaned_df[cleaned_df.UnitPrice == 0.001].index, inplace=True)
cleaned_df.UnitPrice.sort_values()

301033        0.04
221726        0.04
363674        0.04
410140        0.04
407634        0.04
            ...   
268028     4287.63
297723     4505.17
173382     8142.75
299982    11062.06
15017     13541.33
Name: UnitPrice, Length: 524874, dtype: float64

## 保存清理后的数据

完成数据清洗后，把干净整齐的数据保存到新的文件中，文件名为`e_commerce_cleaned.csv`。

In [33]:
cleaned_df.to_csv('e_commerce_cleaned.csv', index=False)
pd.read_csv('e_commerce_cleaned.csv')

C:\Users\stube\AppData\Local\Temp\ipykernel_20552\1651308469.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('e_commerce_cleaned.csv')


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
524869    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
524870    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
524871    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
524872    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
524873    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

                InvoiceDate  UnitPrice CustomerID         Country  
0       2010-12-01 08:26:00       2.55      17850  United Kingdom  
1       2010-12-01 08:26:00       3.39      17850  United Kingdom  
2       2010-12-01 08:26:00       2.75      17850  United Kingdom  
3       2010-12-01 08:26:00       3.39      17850  United Kingdom  
4       2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                     ...        ...        ...             ...  
524869  2011-12-09 12:50:00       0.85      12680          France  
524870  2011-12-09 12:50:00       2.10      12680          France  
524871  2011-12-09 12:50:00       4.15      12680          France  
524872  2011-12-09 12:50:00       4.15      12680          France  
524873  2011-12-09 12:50:00       4.95      12680          France  

[524874 rows x 8 columns]

输出df_cleaned时，有警告信息，同参考版一致